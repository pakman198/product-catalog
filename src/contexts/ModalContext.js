import React, { useState } from 'react';

export const ModalContext = React.createContext();

export const ModalProvider = (props) => {
  const [ isModalVisible, setIsModalVisible ] = useState(false);
  const [ currentItem, setCurrentItem ] = useState({});
  

  const toggleModal = (item) => {
    const selectedItem = item ? item : null;

    setIsModalVisible(!isModalVisible);
    setCurrentItem(selectedItem);
  }

  const contextValue = {
    currentItem,
    isModalVisible,
    toggleModal
  }
    
  return (
    <ModalContext.Provider value={contextValue}>
      { props.children }
    </ModalContext.Provider>
  );
}

