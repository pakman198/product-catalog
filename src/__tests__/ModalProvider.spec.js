import React, { useContext } from 'react';
import { 
  render, 
  waitForElement, 
  fireEvent,
} from '@testing-library/react'
import '@testing-library/jest-dom';

import { ModalContext, ModalProvider } from '../contexts/ModalContext';

import data from '../__mocks__/apiResponse.json';

describe('ModalProvider', () => {


  const renderComponent = (item = null) => {
    return render(
      <ModalProvider>
        <ModalContext.Consumer>
        {context => (
          <div>
            <button onClick={() => context.toggleModal(item)}>Click Me</button>
            <span data-testid="item">{ context.currentItem ? 'Item' : 'Select an Item' }</span>
          </div>
        )}
      </ModalContext.Consumer>
      </ModalProvider>
    )
  }

  test('It renders properly', () => {
    const { asFragment } = renderComponent(data[0]);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Set Context.currentItem when Context.toggleModal function is called', async () => {
    const { getByText, getByTestId } = renderComponent(data[0]);
    const button = await waitForElement(() => getByText('Click Me'));
    fireEvent.click(button);

    const item = await waitForElement(() => getByTestId('item'));
    
    expect(item.innerHTML).toBe('Item');
  });

  test('Set Context.currentItem to null when Context.toggleModal function is called without a parameter', async () => {
    const { getByText, getByTestId } = renderComponent();
    const button = await waitForElement(() => getByText('Click Me'));
    fireEvent.click(button);

    const item = await waitForElement(() => getByTestId('item'));
    
    expect(item.innerHTML).toBe('Select an Item');
  });
});