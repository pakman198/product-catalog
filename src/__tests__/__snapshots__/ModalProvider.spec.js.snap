// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalProvider It renders properly 1`] = `
<DocumentFragment>
  <div>
    <button>
      Click Me
    </button>
    <span
      data-testid="item"
    >
      Item
    </span>
  </div>
</DocumentFragment>
`;
